### ISSUE TITLE
Discrepancy between training and evaluation energies

### ISSUE THREAD

**Issue by mmezic** [deepqmc/deepqmc#97]
There is a mismatch between the training energy and the evaluation energy returned by PauliNet. The mismatch is observable only for larger atoms (I've observed it only for the 4th row of the periodic table so far). Here is an example of the issue on an atom of Ti. The plot depicts the iteration energy as well as its exponentially weighted moving average. First, I train the network, then I evaluate it for a certain number of steps (i.e. without training, passing `opt=None`). By averaging over the evaluation iterations we should get a more accurate energy estimate. We would expect the energy to fluctuate around a constant value throughout the evaluation process. However, in reality the average energy seems to asymptotically approach a higher energy. This value is considerably higher than the last few training energies.

For the sake of interest, I continued training after the evaluation period. You can see that the training energy picks up at roughly the same value where the first training period left off. Note that I got qualitatively similar results when I repeated the process with different seeds, although the energy mismatch is not always as pronounced.
![image](https://user-images.githubusercontent.com/18701830/217509836-325a2a1d-8eba-41f7-9bb7-c77c4bfae8d1.png)

I don't know what's causing this issue, but it looks like there might be some non-negligible undersampled regions of the wavefunction (maybe even spurious modes?). I think that when these regions are encountered during evaluation, their energy contribution increases the total energy without being quickly suppressed, unlike during the training process.

Unfortunately, this issue persists even after the introduction of pseudopotentials (see pull request #83), but the energy discrepancy is much smaller and I had to use more training steps to even observe the issue. Although the average evaluation energy is probably still within the uncertainty range of the training energy, I consistently observe higher evaluation energies when testing my pseudopotential code. And if we benchmark the PauliNet accuracy, even a small increase in energy becomes very important. In this case, only 12 outermost electrons are explicitly treated. This suggests that the issue arises due to the complicated nature of heavy atoms rather than the large number of interacting electrons.
![image](https://user-images.githubusercontent.com/18701830/217510736-6afd4290-3bee-45c7-811a-f4de5308fdb5.png)

**Comment by szbernat** [comment]
This is super interesting.
Since the code that performs the training and evaluation is practically the same, I think it's unlikely that this is caused by some bug that only affects one but not the other (but let's also keep this possibility in the backs of our minds).
A quick sanity check would be to try and find some reference data for the Ti atom, since the difference there is several hartrees, I think we would be able to tell whether the training or the evaluation matches it better.

Again since everything is so similar between training and evaluation I can only think of two causes right now. Investigating these will require many experiments, so before we begin it might be a good idea to try and come up with a calculation that still produces this behaviour and is as cheap to perform as possible. For example: reduce the system size, reduce the network size, reduce the number of initial training iterations (maybe start from some checkpoint). Then we could start investigating the following things a lot more efficiently:
1. The KFAC optimizer does something we don't expect, which changes the local energies. To investigate this we could:
    1. Try using the `adamw` optimizer, and see if the problem persists. A potential issue might be that `adamw` is not be able to find the same solution that KFAC does.
    2. Try using the KFAC optimizer but restrict it to make steps of size zero. You can do this by passing by setting the learning rate and the `norm_constraint` parameters to 0 [here](https://github.com/deepqmc/deepqmc/blob/master/src/deepqmc/conf/task/opt/kfac.yaml).
2. The samples do not equilibrate properly between two training steps, causing the model to overfit to certain regions. I think this is the most likely scenario, although I'm a slightly surprised how quickly the training energies pick up where they left off after an evaluation period. To investigate this we should try to ensure that the samples do have time to equilibrate. I can think of a few avenues to try (individually and even simultaneously):
    1. Decrease the learning rate of the optimizer. Such that it cannot change the WF as much in as single iteration, and the samples can equilibrate more easily. Maybe we could start with setting the learning rate to zero, to see if we are on the right track.
    2. Try to use the `LangevinSampler` instead of the `MetropolisSampler`, for which the average step size (`tau` value) is usually larger.
    3.  Increase the decorrelation lengths of the `DecorrSampler`. We might have to go crazy here, I can imagine values of a few hundred instead of the default 20.
    4. Again, use the `adam` or `adamw` optimizers, perhaps those cannot overfit so easily.
    5. Use a larger `sample_size`. With more independent samples, it should be more difficult to overfit. A potential issue is the memory/computational bottleneck.

It would be great if we could observe some effect of one or a combination of multiple of these things. Then we could determine new defaults that we can use for these heavier elements. If none of these solve the issue, then there must be something else at play.

**Comment by mmezic** [comment]
Thanks for the suggestions. 
 
I have actually also tried running the evaluation period with the kfac optimiser whose learning rate was set to 0 to check if there were any bugs in the evaluation code. I got essentially the same result as with `opt=None`.

Here is the second plot with a reference energy taken from this [FermiNet paper](https://journals.aps.org/prresearch/abstract/10.1103/PhysRevResearch.4.013021). The training energy matches better. Unfortunately, I haven't found any reference data for the case without pseudopotential.
![image](https://user-images.githubusercontent.com/18701830/217775555-f36c611e-d9d5-4a77-8b45-eebb2e5071bc.png)

The fact that this issue only arises for large atoms makes it difficult to investigate (this plot took about 2 days to compute). For example, I've seen a slight hint of the energy mismatch even for an atom of S, but the large variance makes it very hard to distinguish between the stochasticity and an actual energy shift (it could be resolved by using more samples, which defeats the benefit of the smaller system). Nevertheless, I'll try to reduce this issue to as cheap a computation as possible, and try to narrow down the possible causes of the issue based on your suggestions. I think your 2nd remark is the case, and the problem lies somewhere in the sampling.

**Comment by mmezic** [comment]
Hey, I just wanted to post an update here as we now have more knowledge about the energy discrepancy issues. It is quite tedious to investigate, because the energy jumps only occur in large systems. I would like to summarise here what the problems are, and what I think is going on. I think there are at least three different effects at play.

## 1) Correlation of samples
This effect should be more or less clear from my last comment. When the decorrelation length is too short, the model will overfit on these correlated samples and will reach unphysical lower energies. The evaluation will then show the correct energy of an ansatz, which is higher.

I'd like to thank Alice, who ran similar experiments to confirm this.

## 2) Small bidirectional energy jumps for large systems
For large systems, the evaluation energy is sometimes slightly different from the last training energy. In some cases it's larger, in others, it's smaller. The size (and direction) of the energy jump also depends on the checkpoint we are evaluating. Even one training step can cause a noticeable difference in the evaluation energy. Although these differences are tiny, they become significant when trying to achieve high accuracy.

The following plot shows the training (dark line) and several subsequent evaluations (colors) from some of the last checkpoints separated by individual training steps.
<img width="354" alt="Cr_example_jumps" src="https://github.com/deepqmc/deepqmc/assets/18701830/d0bca510-2477-49c8-a3cf-54534a23b5ff">

My guess is that these jumps are a natural manifestation of a learning rate that is too large and thus the energy simply fluctuates too much during the training. When I set a very small learning rate (1000x smaller than the default), the jumps disappeared. However, I'll have to run more experiments to confirm that this is the case. I'd also like to do a comparison of the evaluation energies and their variances to see if they are actually inconsistent.

## 3) Inconsistencies between evaluation and training with zero learning rate (lr)

When I mentioned evaluation above, I meant training with lr set to zero. If I use the actual evaluation instead (i.e. pass the argument `opt=None`), I get a different issue, although it is supposed to be the same as training with `lr=0`. It's very similar to problem 2), but the energy differences are much larger. They are definitely outside the variance and couldn't be explained by too high learning rate. The following plot shows an example of this issue. It depicts the training curve of the TiO molecule with several evaluations from checkpoints saved during the training period.

<img width="354" alt="TiO_eval_dis_unnoted" src="https://github.com/deepqmc/deepqmc/assets/18701830/df215c97-8ba3-47d0-b47c-ce89fb715dc2">

I just recently noticed that I only see this different behavior of evaluation and training with `lr=0` only on a cluster with A100 GPUs. When I run the same codes on RTX3090 or GTX1080, the evaluation produces the same results as training with `lr=0`. I am going to look into this more closely and check the versions of my libraries.

**Comment by mmezic** [comment]
I am posting an update on this issue as I have made some progress recently. I've been having problems with the discrepancy between training and evaluation energies when evaluating some of the transition metal oxides. It turned out that it wasn't related to any of the three effects mentioned in my previous post, because
1) I used 100 decorrelation steps, which is much more than we used before. I could have tried to confirm that it was indeed a sufficient number of steps by computing sample autocorrelation, but it still couldn't explain the negative energy jumps.
2) The discrepancies are not actual energy fluctuations caused by too high a learning rate, because I was able to get rid of those fluctuations (see below).
3) I confirmed that there was really no difference between evaluation mode and training with zero learning rate. Both modes matched perfectly on all the GPUs I tested. They even matched on A100 despite observing a clear difference before, so perhaps the newer code version (or newer jax version) has fixed this issue.

### New results

In my recent results, there was a significant discrepancy in the evaluation energy for the CrO molecule, so I used this system to perform the following tests. I tested it on three different GPUs (A5000, A100 and RTX3090) and surprisingly got different results for different GPUs. 

#### A5000
The following plot shows the local energies in the late training stage. A checkpoint was taken at step 50000 which was evaluated (blue curve). The evaluation energy is more than 30 mHa lower than the average training energy (E_ewm) at the 50kth step. This energy difference is consistent over several multiple independent evaluations or trainings with 0 learning rate from the 50k checkpoint.

![image](https://github.com/deepqmc/deepqmc/assets/18701830/edc9ae8c-8b61-4dbb-a133-f49f6e1ae697)

#### RTX3090
If I run this on the RTX3090, I get a very similar result (shown in the last plot), except that the evaluation energy is now shifted up by about 30 mHa instead of down.

#### A100
This is the result of the exactly the same calculation, but from step 50k the code was run on A100. The first plot is the regular training, the second is the evaluation.
![image](https://github.com/deepqmc/deepqmc/assets/18701830/ccc74ea2-0b6d-45f0-b1de-c8932deb9efd)
![image](https://github.com/deepqmc/deepqmc/assets/18701830/a6c04a31-6a93-4142-81ab-27737146e357)
The evaluation energy is now almost exactly the same as the final mean training energy. This is not a coincidence, I was able to reproduce this result with different checkpoints.

### Explanation

An important thing to note is that all three of the GPUs tested are built with the [Ampere](https://en.wikipedia.org/wiki/Ampere_(microarchitecture)) microarchitecture that supports [TensorFloat-32](https://deeprec.readthedocs.io/en/latest/NVIDIA-TF32.html) (`TF32`).  `TF32` is a 19bit format that has the precision of `float16` and the range of `float32`. It is enabled by default on all these three GPUs and it internally replaces the `float32` datatype that we use to store most of the variables. This "silently" decreases the precision (mantissa) of the calculations from 23 bits to 10 bits (but the range, i.e. the 8bit exponent, remains unaffected). 

This is what happens when I switch off `TF32` and calculate with `float32` precision on RTX3090
![image](https://github.com/deepqmc/deepqmc/assets/18701830/d7d4c3d7-1a87-4f89-be6c-71e9fa66c16b)


Contrary to the default setting (green color), the energy without `TF32` (cyan color) now exactly matched the final mean training energy as in the A100 case. So my conclusion is that the discrepancy in evaluation energy was caused by the low precision of `TF32`. `TF32` can be disabled by setting the environment variable `NVIDIA_TF32_OVERRIDE=0`.

I think the reason why A100 produces the correct energy even with `TF32` is because the code is compiled slightly differently on each GPU. The A100 has more than thrice the memory of the A5000 or RTX3090, so perhaps not all variables were optimized as `TF32` during compilation, which could lead to more accurate results.

### Further notes

Since not using `TF32` (or running on A100) considerably reduces the variance of the local energies during training (compare the right side of the first and second plots), I expect that it might also lead to improvements in convergence speed and stability. I will run some tests to compare training with and without `TF32`.

Interestingly, I observed about a 50% slowdown when using `NVIDIA_TF32_OVERRIDE=0` on the A100, while only a 3% slowdown on the RTX3090.
